File: flutter_app/lib/features/productos/domain/entities/producto.dart
/// Entidad de dominio `Producto` (pura, sin serialización ni dependencias de
/// frameworks). El mapeo JSON vive en la capa `data` (`ProductoModel`).
class Producto {
  final int id;
  final String title;
  final String description;
  final double price;
  final String category;
  final int stock;
  final double rating;
  final String thumbnail;

  const Producto({
    required this.id,
    required this.title,
    this.description = '',
    required this.price,
    this.category = '',
    this.stock = 0,
    this.rating = 0,
    this.thumbnail = '',
  });

  Producto copyWith({
    int? id,
    String? title,
    String? description,
    double? price,
    String? category,
    int? stock,
    double? rating,
    String? thumbnail,
  }) {
    return Producto(
      id: id ?? this.id,
      title: title ?? this.title,
      description: description ?? this.description,
      price: price ?? this.price,
      category: category ?? this.category,
      stock: stock ?? this.stock,
      rating: rating ?? this.rating,
      thumbnail: thumbnail ?? this.thumbnail,
    );
  }

  @override
  bool operator ==(Object other) =>
      identical(this, other) || (other is Producto && other.id == id);

  @override
  int get hashCode => id.hashCode;
}

File: react_native_app/src/features/productos/domain/entities/Producto.ts
// Entidad de dominio `Producto` (pura, sin serialización ni dependencias de
// frameworks). El mapeo JSON vive en la capa `data` (`ProductoModel`).

export interface ProductoProps {
  id: number;
  title: string;
  description?: string;
  price: number;
  category?: string;
  stock?: number;
  rating?: number;
  thumbnail?: string;
}

export class Producto {
  readonly id: number;
  readonly title: string;
  readonly description: string;
  readonly price: number;
  readonly category: string;
  readonly stock: number;
  readonly rating: number;
  readonly thumbnail: string;

  constructor(props: ProductoProps) {
    this.id = props.id;
    this.title = props.title;
    this.description = props.description ?? '';
    this.price = props.price;
    this.category = props.category ?? '';
    this.stock = props.stock ?? 0;
    this.rating = props.rating ?? 0;
    this.thumbnail = props.thumbnail ?? '';
  }

  copyWith(changes: Partial<ProductoProps>): Producto {
    return new Producto({
      id: changes.id ?? this.id,
      title: changes.title ?? this.title,
      description: changes.description ?? this.description,
      price: changes.price ?? this.price,
      category: changes.category ?? this.category,
      stock: changes.stock ?? this.stock,
      rating: changes.rating ?? this.rating,
      thumbnail: changes.thumbnail ?? this.thumbnail,
    });
  }
}
